### PR TITLE
DRILL-6279: UI indicates operators that spilled in-memory data to disk

### PR DESCRIPTION
**Pull request by kkhatua**
1. Detect the presence of an operator that has spilled to disk, in the Operators Overview section of a query's profile page.
2. Introduced API to inject html attributes into a cell. This is used to inject details about the spill in the avg peak and max peak memory usage
3. Javascript is leveraged to detect spilled operators' HTML elements, and inject a symbolic font for indicating the operator as having spilt.
4. Mouse-over effect indicates the average and max number of spill cycles the operator went through. 

For example, at the bottom right of this screenshot, you can see how the spilling operator is indicated, along with the mouse-over information
spilltodisksnapshot

NOTE: Bootstrap library comes with Glyphicons Halfling Fonts and meets the Apache license requirements.
http://glyphicons.com/license#halflingsbootstrap

### CODE REVIEW

**Comment by kkhatua** [comment]
@arina-ielchiieva could you review this? I will be using this commit as the basis for [DRILL-6289](https://issues.apache.org/jira/browse/DRILL-6289)

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java [hunk 1]
```diff
@@ -38,7 +39,16 @@
  * Wrapper class for profiles of ALL operator instances of the same operator type within a major fragment.
  */
 public class OperatorWrapper {
+  //private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(OperatorWrapper.class);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: If logger is not needed, remote it.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java [hunk 2]
```diff
@@ -38,7 +39,16 @@
...
+
+  private static final String HTML_ATTRIB_SPILLS = "spills";
+  private static final String HTML_ATTRIB_CLASS = "class";
+  private static final String HTML_ATTRIB_STYLE = "style";
+  private static final String HTML_ATTRIB_TITLE = "title";
+  private static final DecimalFormat DECIMAL_FORMATTER = new DecimalFormat("#.##");
   private static final String UNKNOWN_OPERATOR = "UNKNOWN_OPERATOR";
+  //-ve value to indicate absence of metric
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: `-ve value`?

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java [hunk 3]
```diff
@@ -190,8 +223,59 @@ public void addSummary(TableBuilder tb, HashMap<String, Long> majorFragmentBusyT
...
+    tb.appendBytes(peakMem.getLeft().getPeakLocalMemoryAllocated(), maxSpillMap);
+  }
+
+  /**
+   * Returns index of Spill Count/Cycle metric
+   * @param operatorType
+   * @return index of spill metric
+   */
+  private int getSpillCycleMetricIndex(CoreOperatorType operatorType) {
+    String metricName = null;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: Initializing with null is redundant.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java [hunk 4]
```diff
@@ -190,8 +223,59 @@ public void addSummary(TableBuilder tb, HashMap<String, Long> majorFragmentBusyT
...
-    tb.appendBytes(Math.round(memSum / size));
-    tb.appendBytes(peakMem.getLeft().getPeakLocalMemoryAllocated());
+
+    //Inject spill-to-disk attributes
+    Map<String, String> avgSpillMap = null;
+    Map<String, String> maxSpillMap = null;
+    if (hasSpilledToDisk) {
+      avgSpillMap = new HashMap<>();
+      //Average SpillCycle
+      float avgSpillCycle = (float) spillCycleSum/size;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: Why float? You are dividing double by int.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java [hunk 5]
```diff
@@ -190,8 +223,59 @@ public void addSummary(TableBuilder tb, HashMap<String, Long> majorFragmentBusyT
...
+      avgSpillMap.put(HTML_ATTRIB_TITLE, DECIMAL_FORMATTER.format(avgSpillCycle) + " Spills on average");
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: I think `s` can be in lower case: `spills on average`.

File: exec/java-exec/src/main/resources/rest/profile/profile.ftl [hunk 6]
```diff
@@ -372,6 +372,16 @@ table.sortable thead .sorting_desc { background-image: url("/static/img/black-de
...
+    //Inject Spilled Tags
+    $(window).on('load', function () {
+      var spillLabel = document.getElementsByClassName("spill-tag");
+      var i;
+      for (i = 0; i < spillLabel.length; i++) {
+        var content = spillLabel[i].innerHTML;
+        spillLabel[i].innerHTML = "<span class=\"glyphicon glyphicon-download-alt\">&nbsp;</span>"+content;
+      }
+    });
+
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/resources/rest/profile/profile.ftl]: Why do you need exec/java-exec/src/main/resources/rest/static/img/popout.png?

Comment (kkhatua) [reply exec/java-exec/src/main/resources/rest/profile/profile.ftl]: It was actually meant as a part of DRILL-6289. Will remove it.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java (hunk 5 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: Agreed. Will fix it.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java (hunk 4 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: Needed the value to be in decimal (formatter will limit it to 2 decimal spaces).

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java (hunk 2 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: Using a negative ("-ve") value constant. Will fix the comment

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java (hunk 1 above)
Comment (kkhatua) [reply exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/OperatorWrapper.java]: Ok. I see a lot of other places where it is commented out. I'll uncomment it with an unused tag for the compler to ignore. Comes very handy in debugging when making changes

**Comment by arina-ielchiieva** [comment]
+1

**Pull request closed without merging**